entity-service-impl/src/test/java/org/hypertrace/entity/detector/GibberishDetector.java

```diff
@@ -0,0 +1,142 @@
+  }
+
+  private double getAvgTransitionProbability(String line, double[][] logProbabilityMatrix) {
+    double logProb = 0d;
+    int transitionCount = 0;
+    List<String> nGram = get2Gram(line);
+    for (String touple : nGram) {
+      logProb +=
+          logProbabilityMatrix[alphabetPositionMap.get(touple.charAt(0))][
+              alphabetPositionMap.get(touple.charAt(1))];
+      transitionCount++;
+    }
+    return Math.exp(logProb / Math.max(transitionCount, 1));
+  }
+
+  private void train() {
+    long startTs = System.currentTimeMillis();
+    int[][] alphabetCouplesMatrix = getAlphaBetCouplesMatrix();
+    logProbabilityMatrix = getLogProbabilityMatrix(alphabetCouplesMatrix);
+
+    List<Double> validDataProbability =
+        getAvgTransitionProbability(trainingDataset.getValidData(), logProbabilityMatrix);
+    List<Double> invalidDataProbability =
+        getAvgTransitionProbability(trainingDataset.getInvalidData(), logProbabilityMatrix);
+
+    double minGood = Collections.min(validDataProbability);
+    double maxBad = Collections.max(invalidDataProbability);
+
+    if (minGood <= maxBad) {
+      throw new AssertionError("cannot create a threshold");
+    }
+    threshold = getThreshold(minGood, maxBad);
+    LOGGER.info(
+        "Time taken to train gibberish detector is {} ms", System.currentTimeMillis() - startTs);
+  }
+
+  // can be overridden for another threshold heuristic implementation
+  protected double getThreshold(double minGood, double maxBad) {
+    return (minGood + maxBad) / 2;
+  }
+
+  private void initializePositionMap() {
+    List<Character> alphabetChars = trainingDataset.getEligibleCharacters();
+    for (int i = 0; i < alphabetChars.size(); i++) {
+      alphabetPositionMap.put(alphabetChars.get(i), i);
+    }
+  }
+
+  private String normalize(String line) {
+    StringBuilder normalizedLine = new StringBuilder();
+    for (char c : line.toLowerCase().toCharArray()) {
+      normalizedLine.append(trainingDataset.getEligibleCharacters().contains(c) ? c : "");
+    }
+    return normalizedLine.toString();
+  }
+
+  private List<String> get2Gram(String line) {
+    int n = 2;
+    String filteredLine = normalize(line);
+    List<String> nGram = new ArrayList<>();
+    for (int start = 0; start < filteredLine.length() - n + 1; start++) {
+      nGram.add(filteredLine.substring(start, start + n));
+    }
+    return nGram;
+  }
+
+  private int[][] getAlphaBetCouplesMatrix() {
+    int[][] counts = createArray(trainingDataset.getEligibleCharacters().size());
+    for (String line : trainingDataset.getTrainingData()) {
+      List<String> nGram = get2Gram(line);
+      for (String touple : nGram) {
+        counts[alphabetPositionMap.get(touple.charAt(0))][
+            alphabetPositionMap.get(touple.charAt(1))]++;
+      }
+    }
+    return counts;
+  }
+
+  private double[][] getLogProbabilityMatrix(int[][] alphabetCouplesMatrix) {
```

Comment:
Accidental check-in along with this PR? Or, trying to piggy-back 2 changes in a single PR?